### PR TITLE
Add more v3 tests

### PR DESCRIPTION
**Pull request by ysthakur**
v3 sorely needs more tests. This is just a draft for now.

### CODE REVIEW

**Comment by Lyxal** [comment]
That's right. I'm nowhere near done yet with adding tests for everything currently implemented.

File: shared/src/test/scala/ElementTests.scala
Comment (ysthakur) [inline shared/src/test/scala/ElementTests.scala:214]:
```suggestion
```
Silly billy

File: shared/src/main/scala/Context.scala
Comment (ysthakur) [inline shared/src/main/scala/Context.scala:32]: Great idea making them more visually distinctive. Do we want to call them `ctxVarPrimary/Secondary` so it's a little shorter?

File: shared/src/main/scala/Elements.scala
Comment (ysthakur) [inline shared/src/main/scala/Elements.scala:192]:
```suggestion
    val toBinary = addVect(
```

File: shared/src/main/scala/Context.scala
Comment (Lyxal) [reply shared/src/main/scala/Context.scala:32]: I think that would be fine

File: shared/src/main/scala/Elements.scala
Comment (ysthakur) [inline shared/src/main/scala/Elements.scala:204]:
```suggestion
          a.map(x => NumberHelpers.toBinary(StringHelpers.chrord(x.toString)))*
```

Comment (Lyxal) [reply shared/src/main/scala/Elements.scala:204]: I could have sworn I tried something like that last night and couldn't get it to compile. Does this compile and mean that I'm just bad at scala? :p

File: shared/src/main/scala/Elements.scala
Comment (ysthakur) [inline shared/src/main/scala/Elements.scala:450]:
```suggestion
          if temp.forall(_.isInstanceOf[String])
          then temp.mkString
```

**Review by Lyxal — APPROVED** [review]
An approval for merging when ready. Obviously, don't merge straight away.

File: shared/src/main/scala/Elements.scala
Comment (ysthakur) [inline shared/src/main/scala/Elements.scala:489]: May I ask why dropRight is needed instead of the b-1 from before?

Comment (Lyxal) [reply shared/src/main/scala/Elements.scala:489]: Because of cases like `5 -> 1`

Comment (Lyxal) [reply shared/src/main/scala/Elements.scala:489]: A range from 5 to 0 isn't what should be returned. So instead of having conditional swapping, just remove the last item of the range. That works for ranges in both directions.

File: shared/src/main/scala/NumberHelpers.scala
Comment (ysthakur) [inline shared/src/main/scala/NumberHelpers.scala:58]: Since we're working with 32-bit ints here, is it necessary to use `VNum()`?

File: shared/src/main/scala/NumberHelpers.scala
Comment (ysthakur) [inline shared/src/main/scala/NumberHelpers.scala:78]:
```suggestion
          temp.vmap(v => -v.asInstanceOf[VNum])
```

File: shared/src/main/scala/NumberHelpers.scala
Comment (Lyxal) [reply shared/src/main/scala/NumberHelpers.scala:58]: Yes. It really is. They're cast to VNum for the `%` behaviour.

Default scala modulo doesn't support floating point numbers and is truncated. We need floored floating point modulo

**Review by ysthakur — COMMENTED** [review]
I'd approve this if I could lol

**Comment by Lyxal** [comment]
> I'd approve this if I could lol

That's why I clicked approve lol

File: shared/src/main/scala/Elements.scala
Comment (ysthakur) [reply shared/src/main/scala/Elements.scala:485]: Oh, smart

File: shared/src/main/scala/NumberHelpers.scala
Comment (ysthakur) [reply shared/src/main/scala/NumberHelpers.scala:58]: Ah, right. In that case, we might as well make `current` and the rest `VNum`s too, so we're not restricted to 32-bit ints. `b.toIntegral` should work for that (although we should only call it once and store it in a variable)

**Comment by Lyxal** [comment]
Mmmm compiler errors.

File: shared/src/main/scala/NumberHelpers.scala
Comment (ysthakur) [reply shared/src/main/scala/NumberHelpers.scala:58]: Maybe something like
```scala
  def multiplicity(a: VNum, b: VNum): VNum =
    if a == VNum(0) || b == VNum(0) then return VNum(0)
    if b.vabs == VNum(1) then return a.vabs
    var result = 0
    var current = a.toIntegral
    val bInt = b.toIntegral
    while current % bInt == VNum(0) do
      result += 1
      current /= bInt
    result
```

File: shared/src/main/scala/NumberHelpers.scala
Comment (ysthakur) [inline shared/src/main/scala/NumberHelpers.scala:70]:
```suggestion
  def toBinary(a: VAny)(using Context): VAny =
```

**Comment by ysthakur** [comment]
My bad, I just wanted to use `vmap` to make it more concise

File: shared/src/main/scala/NumberHelpers.scala
Comment (Lyxal) [reply shared/src/main/scala/NumberHelpers.scala:58]: Actually, I think nothing should be cast to int. I was reading the current v2 implementation and it doesn't have any cast to int

Comment (Lyxal) [reply shared/src/main/scala/NumberHelpers.scala:58]: Just leave them all as `VNum`. No need to uncast. I probably only keep the to integral because I was keeping some of the original code from before

Comment (ysthakur) [reply shared/src/main/scala/NumberHelpers.scala:58]: So maybe something like this?
```scala
  def multiplicity(a: VNum, b: VNum): VNum =
    if a == VNum(0) || b == VNum(0) then return VNum(0)
    if b.vabs == VNum(1) then return a.vabs
    var result = 0
    var current = a
    while current % b == VNum(0) do
      result += 1
      current /= b
    result
```

Comment (Lyxal) [reply shared/src/main/scala/NumberHelpers.scala:58]: Exactly like that

**Pull request merged**
